### PR TITLE
[x-usage] add x-usage command

### PR DESCRIPTION
**Pull request by strega-nil**
This allows one to get the usage of a package in manifest mode.

### CODE REVIEW

**Comment by ras0219** [comment]
This doesn't solve the fundamental problem which is that the install dir is not well defined in manifest mode.

**Comment by autoantwort** [comment]
Imo this command should be added here because it is very useful for users:

https://github.com/microsoft/vcpkg-tool/blob/eea14d0feb2b26d36031405af42e1e48199e4168/src/vcpkg/vcpkgcmdarguments.cpp#L557

**Comment by autoantwort** [comment]
> This doesn't solve the fundamental problem which is that the install dir is not well defined in manifest mode.

But do you need the install dir? What packages are requested/installed is defined in the vcpkg.json file in manifest mode.

**Comment by ras0219** [comment]
>What packages are requested/installed is defined in the vcpkg.json file in manifest mode.

The vcpkg packages are known, but the results of those packages are unknown until after their installed. vcpkg's usage information is derived by heuristically scanning the results of the package for `-config.cmake` files, which means we need the install dir (with installed packages) to display the information.

**Comment by JackBoosY** [comment]
Fixes https://github.com/microsoft/vcpkg/issues/15242.

**Comment by autoantwort** [comment]
> vcpkg's usage information is derived by heuristically scanning the results of the package for `-config.cmake` files, which means we need the install dir (with installed packages) to display the information.

Maybe write the usage information to a file next to the vcpkg.json file while installing the dependencies in manifest mode.

File: src/vcpkg/commands.usage.cpp [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        auto result = Json::Object{};
+        if (const std::string* usage_file = cmui.usage_file.get())
+        {
+            result.insert("usage-file", Json::Value::string(*usage_file));
+        }
+        else if (const std::string* header_path = cmui.header_to_find.get())
+        {
+            result.insert("header-to-find", Json::Value::string(*header_path));
+        }
+        else if (!cmui.cmake_targets_map.empty())
```
Comment (BillyONeal) [inline src/vcpkg/commands.usage.cpp:94]: negated if followed by else pet peeve

Comment (BillyONeal) [reply src/vcpkg/commands.usage.cpp:94]: Ditto below

**Review by BillyONeal — APPROVED** [review]

File: src/vcpkg/commands.usage.cpp (hunk 1 above)
Comment (strega-nil-ms) [reply src/vcpkg/commands.usage.cpp:94]: I disagree with this, since what this is actually checking is `cmui.cmake_targets_map.has_value()` or w/e

Comment (BillyONeal) [reply src/vcpkg/commands.usage.cpp:94]: It still forces a reader to negate a negation in order to figure out what the invariants are in the else.

**Comment by ras0219-msft** [comment]
>This doesn't solve the fundamental problem which is that the install dir is not well known in manifest mode outside the CMake invocation.

@strega-nil could you comment on how this issue is solved and describe the specific sequence of user actions that this enables?

**Comment by strega-nil** [comment]
@ras0219-msft I think the most important thing for me here is to enable us to script things in and grab the info. Yes, you may not be calling `x-usage` directly, but we can add, say, a variable to `vcpkg.cmake` that allows us to print usage.

**Comment by ras0219-msft** [comment]
We do already have a semi-scriptable solution for this:

```
PS C:\src\vcpkg> ./vcpkg x-package-info --x-json --x-installed zlib:x86-windows
{
  "results": {
    "zlib:x86-windows": {
      "version-string": "1.2.11",
      "port-version": 10,
      "triplet": "x86-windows",
      "abi": "23bb2ab4c3b9c2a5c87f3f37c78d230dd587dc22",
      "usage": "The package zlib is compatible with built-in CMake targets:\r\n\r\n    find_package(ZLIB REQUIRED)\r\n    target_link_libraries(main PRIVATE ZLIB::ZLIB)\r\n\n",
      "owns": [
        "x86-windows/",
        "x86-windows/bin/",
        "x86-windows/bin/zlib.pdb",
        "x86-windows/bin/zlib1.dll",
        "x86-windows/debug/",
        "x86-windows/debug/bin/",
        "x86-windows/debug/bin/zlibd.pdb",
        "x86-windows/debug/bin/zlibd1.dll",
        "x86-windows/debug/lib/",
        "x86-windows/debug/lib/pkgconfig/",
        "x86-windows/debug/lib/pkgconfig/zlib.pc",
        "x86-windows/debug/lib/zlibd.lib",
        "x86-windows/include/",
        "x86-windows/include/zconf.h",
        "x86-windows/include/zlib.h",
        "x86-windows/lib/",
        "x86-windows/lib/pkgconfig/",
        "x86-windows/lib/pkgconfig/zlib.pc",
        "x86-windows/lib/zlib.lib",
        "x86-windows/share/",
        "x86-windows/share/zlib/",
        "x86-windows/share/zlib/copyright",
        "x86-windows/share/zlib/usage",
        "x86-windows/share/zlib/vcpkg_abi_info.txt"
      ]
    }
  }
}
```

The `"usage"` field reports the data we would have printed to the command line.

Another example of just the usage field for an autodetected library instead of a `usage` file:
```
"usage": "The package abseil:x86-windows provides CMake targets:\n\n    find_package(absl CONFIG REQUIRED)\n    # Note: 123 target(s) were omitted.\n    target_link_libraries(main PRIVATE absl::any absl::base absl::bits absl::city)\n\n",
```

**Comment by ras0219-msft** [comment]
Moving to draft since no progress past July

**Pull request closed without merging**